Add git workflow doc for issue #316

hungry1526 opened the pull request:
Fix #316

hungry1526 committed:
Add git workflow doc

hungry1526 committed:
Update git_workflow.md

tonyyang-svail commented on doc/git_workflow.md:
```diff
@@ -0,0 +1,41 @@
+# Git Workflow for SQLFlow project
+This doc is mainly written for new contributors in this project, those who are experienced open source developers can skip the doc. Please read carefully and leave comments as it needs and helps us improve as more people from the community are joing force. 
```
"joing force" => "joining forces"?

hungry1526 replied:
Will fix.

tonyyang-svail commented on doc/git_workflow.md:
````diff
@@ -0,0 +1,41 @@
+## Fetch Upstream
+First let's run git fetch to download commits, files, and refs from a remote repository into your local repo
+```bash
+git fetch upstream develop
````
You may need to clarify `upstream` and `origin`. For example, if one `go get github.com/sql-machine-learning/sqlflow`. In his local repo, the `origin` refers to the official repo https://github.com/sql-machine-learning/sqlflow, instead of his fork.

Since our `build.md` recommends `go get` to download the source code, I would suggest adjusting `upstream` and `origin` accordingly. Also, adding a sample of `.git/config` will also help. Here is mine:
```
$ cat .git/config
[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
	ignorecase = true
	precomposeunicode = true
[remote "origin"]
	url = https://github.com/sql-machine-learning/sqlflow
	fetch = +refs/heads/*:refs/remotes/origin/*
	fetch = +refs/pull/*/head:refs/remotes/pr/*
[remote "tonyyang-svail"]
	url = https://github.com/tonyyang-svail/sqlflow
	fetch = +refs/heads/*:refs/remotes/origin/*
	fetch = +refs/pull/*/head:refs/remotes/pr/*
```

hungry1526 replied:
Sure, will use your git config as an example here. Sounds good?

hungry1526 replied:
Resolved and committed in the PR, please review.

hungry1526 committed:
Update git_workflow.md with comments from Tony

wangkuiyi commented on doc/git_workflow.md:
````diff
@@ -0,0 +1,63 @@
+
+First let's define upstream and origin branches. In this example, the origin means the sql-machine-learning/sqlflow the original repo whereas upstream means the forked one from sql-machine-learning/sqlflow. Here is my setup:
+
+```bash
````
Please be aware that everything in the `.git` subdirectory is subject to change with the version of Git. It might not be a good idea to reveal the details like `.git/config` file.

hungry1526 replied:
Good idea, will make a change.

wangkuiyi commented on doc/git_workflow.md:
````diff
@@ -0,0 +1,63 @@
+## Checkout Remote Develop for the latest code
+Then we checkout remote develop branch which contains the latest code and commits. This command make sure we start from the most updated state of the repo.
+```bash
+git checkout remotes/origin/develop
````
Usually, developers don't need to type such a command. When we do `git clone`, our local repo would be set to track the default branch of the remote repo.

hungry1526 replied:
Do you mean, when people do ```git clone```, remote branch will be checked out automatically? If so I will add a note here, here this command is to explicitly check out remote branch and make sure we have the latest code.

wangkuiyi commented on doc/git_workflow.md:
````diff
@@ -0,0 +1,63 @@
+Lastly, you may clean up local feature branch using below command. Note, -D will force delete for unmerged changes in the local branch. 
+
+```bash
+git branch -D new_feature_branch_issue_000
````
It's a good working style to clean up. To make it cleaner, we also need to delete the remote branch:

```bash
git push origin :new_feature_branch_issue_000
```

hungry1526 replied:
Sure, will add clean up for remote branch.

wangkuiyi reviewed (commented):
This is a great learning note. I rarely see people tracking into the configuration file in the `.git` subdirectory.

It might be too much for new Git and Github users.

I agree that we need to explain our Git workflow somewhere in our documentation, either adding a new Markdown file like in this PR, or adding it to the `doc/contribution.md`.

I feel that we should present two parts:

1. For experienced contributors, we need to tell that we are following this branching model https://nvie.com/posts/a-successful-git-branching-model/.

1. For new Git and Github users, we need to present a sequence of commands:

   1. `go get sqlflow.org/sqlflow`, which runs `git clone https://github.com/sql-machine-learning/sqlflow/` and other cloning commands for all dependencies.
   1. `git remote rename origin upstream` and `git remote add origin https://github.com/me/sqlflow`. Here we need to explain `upstream` and `origin`; we can simply paste a link https://help.github.com/en/articles/fork-a-repo, which also explains the *fork* operation on Github.
   1. `git checkout -b  my_feature`
   1. `git commit` and `git push -u origin my_feature`
   1. Create pull request
   1. Clean up by `git checkout develop && git pull && git branch -d my_feature && git push origin :my_feature`.

This repo should be focusing on SQLFlow itself.  I would highly appreciate it if @hungry1526 can create a tutorial on Git and Github for beginners. If so, we could put a link in the SQLFlow document pointing to the tutorial hosted in @hungry1526 's Github account.

hungry1526 commented:
@wangkuiyi Good suggestion, will move this file  to my own repo.

hungry1526 committed:
Update git_workflow.md

hungry1526 commented:
@hungry1526 @tonyyang-svail , all the comments are resolved and updated in my personal repo: https://github.com/hungry1526/SQLFlow/blob/develop/doc/git_workflow.md

Contribution section in Readme.md is also updated with new pointer to the doc. Please link to the origin wherever appropriate. 

Thanks for the review and suggestions.
